### PR TITLE
Fix ARCSequenceOpts bottom up handling of terminators 

### PR DESCRIPTION
**Pull request by meg-gupta**
Historically TermInsts were handled specially while visiting blocks in bottom up order. TermInsts were not visited while traversing the block which they belong to. They were visited while traversing successors, and the most conservative terminator of all predecessors would affect the refcount state in the dataflow.

This was needed because ARCSequenceOpts also computed 'insertion points' for the purposes of ARC code motion. ARC code motion was then removed from ARCSequenceOpts and this code remained unchanged.

With this change, arc significant terminators are handled like all other instructions while processing basic blocks bottom up.

Also updateForSameLoopInst, updateForDifferentLoopInst, updateForPredTerminators all serve similar purpose with subtle differences. This change removes some code duplication due to this.

Why this change now ? #33267 improved guaranteed args handling in updateForSameLoopInst. It did not updateForPredTerminators, causing different top down and bottom up behavior for arc significant terminators.

Fixes rdar://66932268

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci test

File: lib/SILOptimizer/ARC/GlobalARCSequenceDataflow.cpp [hunk 1]
```diff
@@ -254,8 +222,17 @@ bool ARCSequenceDataflowEvaluator::processBBBottomUp(
       RCIA, EAFI, BBState, FreezeOwnedArgEpilogueReleases, IncToDecStateMap,
       SetFactory);
 
-  // For each terminator instruction I in BB visited in reverse...
-  for (auto II = std::next(BB.rbegin()), IE = BB.rend(); II != IE;) {
+  auto II = BB.rbegin();
+  if (isa<TermInst>(*II)) {
+    auto *terminator = &cast<TermInst>(*II);
+    if (terminator == BB.getTerminator() &&
+        !isARCSignificantTerminator(terminator)) {
```
Comment (atrick) [inline lib/SILOptimizer/ARC/GlobalARCSequenceDataflow.cpp:229]: It seems like there's some extra logic here. Why not just
```
if (!isARCSignificantTerminator(cast<TermInt>(*II))
 ++II;
```

File: lib/SILOptimizer/ARC/ARCSequenceOptUtils.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+#ifndef SWIFT_SILOPTIMIZER_ARC_ARCSEQUENCEOPTUTILS_H
+#define SWIFT_SILOPTIMIZER_ARC_ARCSEQUENCEOPTUTILS_H
+
+#include "swift/SIL/SILInstruction.h"
+
+namespace swift {
+  bool isARCSignificantTerminator(TermInst *TI);
+} // end namespace swift
+
+#endif
```
Comment (atrick) [inline lib/SILOptimizer/ARC/ARCSequenceOptUtils.h:27]: You may want to insert newlines at the end of the new files so tools don't warn us about that.

**Review by atrick — APPROVED** [review]
Wow! This is a lot better. Thanks.

Just a couple comments.

File: lib/SILOptimizer/ARC/GlobalARCSequenceDataflow.cpp (hunk 1 above)
Comment (meg-gupta) [reply lib/SILOptimizer/ARC/GlobalARCSequenceDataflow.cpp]: Done.

File: lib/SILOptimizer/ARC/ARCSequenceOptUtils.h (hunk 2 above)
Comment (meg-gupta) [reply lib/SILOptimizer/ARC/ARCSequenceOptUtils.h]: Done

**Comment by meg-gupta** [comment]
@swift-ci smoke test

**Comment by meg-gupta** [comment]
@swift-ci benchmark

**Comment by meg-gupta** [comment]
@swift-ci please benchmark

**Review by gottesmm — COMMENTED** [review]
Thanks for eliminating the references to insertion points! This code is really crusty and needed attention = ).

**Comment by meg-gupta** [comment]
@gottesmm thanks for the review. I'll resolve your comments in another PR

**Pull request merged**